File: data_reports.py
#!/usr/bin/env python3

import json
from datetime import datetime

import flask
import pandas as pd
import sqlalchemy
from sqlalchemy.sql.functions import coalesce
from sqlalchemy.sql.operators import and_

import contract_report_model as model
import db_utils
import token_utils
import base64

data_reports_app = flask.Blueprint("data_reports_app", __name__)


@data_reports_app.route("/data-reports/<id>", methods=["PUT"])
def get_reports(id):
    try:
        role = token_utils.get_role()
    except RuntimeError as e:
        return flask.Response(
            str(e),
            status=400,
        )
    try:
        token_utils.check_role(
            [
                db_utils.Role.TEACHER,
                db_utils.Role.HEAD_OF_CATHEDRA,
                db_utils.Role.INSPECTOR,
                db_utils.Role.HEAD_OF_HUMAN_RESOURCES,
            ],
            role,
        )
    except RuntimeError as e:
        return flask.Response(
            str(e),
            status=400,
        )

    if not id in [
        db_utils.ReportTypes.SIGNING_PROGRESS.value,
        db_utils.ReportTypes.EXECUTION_PROGRESS.value,
        db_utils.ReportTypes.SIGNING_LOG.value,
        db_utils.ReportTypes.DATA_CHANGE_LOG.value,
    ]:
        return flask.Response(
            "Unknown id for data report selected",
            status=400,
        )
    data = flask.request.get_json()
    error = ""
    if data.get("from") is None:
        error += "Empty from. "
    if data.get("till") is None:
        error += "Empty till. "
    if data.get("extended") is None:
        error += "Empty extended. "

    if error:
        return flask.Response(
            error,
            status=400,
        )

    date_from = datetime.strptime(data["from"], "%Y-%m-%d").date()
    date_till = (datetime.strptime(data["till"], "%Y-%m-%d").date(),)

    with db_utils.auto_session(db_utils.engine_reader) as session:
        if id == db_utils.ReportTypes.EXECUTION_PROGRESS.value:
            contracts_tmp = (
                session.query(model.Contracts.id.label("contract_id"))
                .select_from(model.Contracts)
                .join(model.Users, model.Users.id == model.Contracts.user_id)
                .join(model.Cathedras, model.Cathedras.id == model.Users.cathedra_id)
                .filter(
                    sqlalchemy.sql.operators.and_(
                        model.Contracts.valid_from <= date_from,
                        model.Contracts.valid_till >= date_till,
                    )
                )
            )

            if data.get("user"):
                contracts_tmp = contracts_tmp.filter(
                    model.Users.id == data["user"]["id"]
                )
            elif data.get("cathedra"):
                contracts_tmp = contracts_tmp.filter(
                    model.Users.cathedra_id == int(data["cathedra"]["id"])
                )
            elif data.get("faculty"):
                contracts_tmp = contracts_tmp.filter(
                    model.Cathedras.faculty_id == int(data["faculty"]["id"])
                )

            contracts_tmp = contracts_tmp.cte("contracts_tmp")

            done_params = (
                session.query(
                    contracts_tmp.c.contract_id,
                    model.ReportedParameters.parameter_id,
                    sqlalchemy.sql.func.sum(model.ReportedParameters.done).label(
                        "done"
                    ),
                )
                .select_from(contracts_tmp)
                .join(
                    model.Reports,
                    sqlalchemy.sql.operators.and_(
                        model.Reports.contract_id == contracts_tmp.c.contract_id,
                        model.Reports.signed_by_head_of_human_resources,
                    ),
                )
                .join(
                    model.ReportedParameters,
                    model.ReportedParameters.report_id == model.Reports.id,
                )
                .group_by(
                    contracts_tmp.c.contract_id, model.ReportedParameters.parameter_id
                )
                .cte("done_params")
            )

            if data["extended"]:
                stmt = sqlalchemy.select(
                    contracts_tmp.c.contract_id,
                    model.ParametersTemplates.parameter_id,
                    model.ParametersTemplates.requirement,
                    model.ParametersTemplates.points_promised,
                    sqlalchemy.sql.func.coalesce(done_params.c.done, 0).label("done"),
                    (
                        sqlalchemy.sql.func.coalesce(done_params.c.done, 0).cast(
                            sqlalchemy.DECIMAL
                        )
                        / model.ParametersTemplates.requirement
                        * 100
                    )
                    .cast(sqlalchemy.Integer)
                    .label("done_perc"),
                    (
                        model.ParametersTemplates.points_promised
                        * sqlalchemy.sql.func.coalesce(done_params.c.done, 0).cast(
                            sqlalchemy.DECIMAL
                        )
                        / model.ParametersTemplates.requirement
                    )
                    .cast(sqlalchemy.Integer)
                    .label("done_points"),
                )
            else:
                stmt = sqlalchemy.select(
                    contracts_tmp.c.contract_id,
                    sqlalchemy.sql.func.sum(
                        model.ParametersTemplates.points_promised
                    ).label("points_promised"),
                    sqlalchemy.sql.func.sum(
                        (
                            model.ParametersTemplates.points_promised
                            * sqlalchemy.sql.func.coalesce(done_params.c.done, 0).cast(
                                sqlalchemy.DECIMAL
                            )
                            / model.ParametersTemplates.requirement
                        ).cast(sqlalchemy.Integer)
                    ).label("done_points"),
                    (
                        100
                        * sqlalchemy.sql.func.sum(
                            (
                                model.ParametersTemplates.points_promised
                                * sqlalchemy.sql.func.coalesce(
                                    done_params.c.done, 0
                                ).cast(sqlalchemy.DECIMAL)
                                / model.ParametersTemplates.requirement
                            ).cast(sqlalchemy.Integer)
                        )
                        / sqlalchemy.sql.func.sum(
                            model.ParametersTemplates.points_promised
                        )
                    )
                    .cast(sqlalchemy.Integer)
                    .label("done_perc"),
                )

            stmt = (
                stmt.select_from(contracts_tmp)
                .join(
                    model.Contracts,
                    model.Contracts.id == contracts_tmp.c.contract_id,
                    isouter=True,
                )
                .join(
                    model.ParametersTemplates,
                    model.ParametersTemplates.template_id
                    == model.Contracts.template_id,
                    isouter=True,
                )
                .join(
                    done_params,
                    sqlalchemy.sql.operators.and_(
                        done_params.c.parameter_id
                        == model.ParametersTemplates.parameter_id,
                        done_params.c.contract_id == contracts_tmp.c.contract_id,
                    ),
                    isouter=True,
                )
            )

            if not data["extended"]:
                stmt = stmt.group_by(contracts_tmp.c.contract_id)

            df = pd.read_sql(stmt, session.bind)

            def color_negative_red(val):
                color = "red" if val < 0 else "black"
                return f"color: {color}"

            df.style.applymap(color_negative_red)
            report_html_body = df.to_html()
            return flask.Response(report_html_body, mimetype="text/xml")
        elif id == db_utils.ReportTypes.SIGNING_PROGRESS.value:

            teachers_table = (
                session.query(
                    model.Contracts.user_id.label("user_id"),
                    model.Reports.id.label("report_id"),
                )
                .select_from(model.Reports)
                .join(model.Contracts, model.Reports.contract_id == model.Contracts.id)
                .filter(model.Reports.period_of_report.between(date_from, date_till))
                .filter(model.Reports.signed_by_teacher == False)
                .cte("teachers_table")
            )

            head_of_cathedra_table = (
                session.query(
                    model.Cathedras.head_id.label("user_id"),
                    model.Reports.id.label("report_id"),
                )
                .select_from(model.Reports)
                .join(model.Contracts, model.Reports.contract_id == model.Contracts.id)
                .join(model.Users, model.Users.id == model.Contracts.user_id)
                .join(model.Cathedras, model.Cathedras.id == model.Users.cathedra_id)
                .join(
                    model.OpenedPeriodForReports,
                    model.OpenedPeriodForReports.period
                    == model.Reports.period_of_report,
                )
                .filter(model.Reports.period_of_report.between(date_from, date_till))
                .filter(model.Reports.signed_by_head_of_cathedra == False)
                .filter(model.OpenedPeriodForReports.time_of_closing < datetime.today())
                .cte("head_of_cathedra_table")
            )

            inspectors_table = (
                session.query(
                    model.Parameters.inspector_id.label("user_id"),
                    model.Reports.id.label("report_id"),
                )
                .select_from(model.Reports)
                .join(
                    model.ReportedParameters,
                    model.ReportedParameters.report_id == model.Reports.id,
                )
                .join(
                    model.Parameters,
                    model.Parameters.id == model.ReportedParameters.parameter_id,
                )
                .join(
                    model.OpenedPeriodForReports,
                    model.OpenedPeriodForReports.period
                    == model.Reports.period_of_report,
                )
                .filter(model.Reports.period_of_report.between(date_from, date_till))
                .filter(model.OpenedPeriodForReports.time_of_closing < datetime.today())
                .filter(model.ReportedParameters.signed_by_inspector == False)
                .cte("inspectors_table")
            )

            stmt_teachers = create_html_table_for_report(teachers_table, data)
            stmt_head_of_cathedra_table = create_html_table_for_report(
                head_of_cathedra_table, data
            )
            stmt_inspectors_table = create_html_table_for_report(inspectors_table, data)

            df_1 = pd.read_sql(stmt_teachers, session.bind)
            df_2 = pd.read_sql(stmt_head_of_cathedra_table, session.bind)
            df_3 = pd.read_sql(stmt_inspectors_table, session.bind)

            def color_negative_red(val):
                color = "red" if val < 0 else "black"
                return f"color: {color}"

            df_1.style.applymap(color_negative_red)
            df_2.style.applymap(color_negative_red)
            df_3.style.applymap(color_negative_red)

            report_html_body = (
                "<label>Teachers:</label>"
                + df_1.to_html()
                + "<label>Head of cathedra:</label>"
                + df_2.to_html()
                + "<label>Inspectors:</label>"
                + df_3.to_html()
            )
            return flask.Response(report_html_body, mimetype="text/plain")
        elif id == db_utils.ReportTypes.SIGNING_LOG.value:

            log_table = (
                sqlalchemy.select(
                    model.SignatureLogs.time_of_change,
                    model.Users.full_name,
                    model.Reports.period_of_report,
                    model.SignatureLogs.action,
                )
                .join(model.SignatureLogs.report)
                .join(model.SignatureLogs.user)
                .where(model.SignatureLogs.time_of_change.between(date_from, date_till))
                .order_by(model.SignatureLogs.time_of_change)
            )

            df = pd.read_sql(log_table, session.bind)

            def color_negative_red(val):
                color = "red" if val < 0 else "black"
                return f"color: {color}"

            df.style.applymap(color_negative_red)

            return flask.Response(df.to_html(), mimetype="text/plain")
        elif id == db_utils.ReportTypes.DATA_CHANGE_LOG.value:

            log_table = (
                sqlalchemy.select(
                    model.DataChangeLogs.time_of_change,
                    model.Users.full_name,
                    model.DataChangeLogs.object_of_change,
                    model.DataChangeLogs.befor_change,
                    model.DataChangeLogs.after_change,
                )
                .join(model.DataChangeLogs.user)
                .where(model.DataChangeLogs.time_of_change.between(date_from, date_till))
                .order_by(model.DataChangeLogs.time_of_change)
            )

            df = pd.read_sql(log_table, session.bind)

            def color_negative_red(val):
                color = "red" if val < 0 else "black"
                return f"color: {color}"

            df.style.applymap(color_negative_red)

            return flask.Response(df.to_html(), mimetype="text/plain")


def create_html_table_for_report(table, data):
    userReport = sqlalchemy.orm.aliased(model.Users)
    stmt = (
        sqlalchemy.select(
            (model.Users.full_name + " (" + table.c.user_id + ")").label("user_id"),
            (
                model.Reports.period_of_report.cast(sqlalchemy.String)
                + " ("
                + userReport.full_name
                + ")"
            ).label("report_id"),
        )
        .select_from(table)
        .join(model.Users, model.Users.id == table.c.user_id)
        .join(model.Reports, model.Reports.id == table.c.report_id)
        .join(model.Contracts, model.Contracts.id == model.Reports.contract_id)
        .join(userReport, userReport.id == model.Contracts.user_id)
        .join(model.Cathedras, model.Cathedras.id == userReport.cathedra_id)
    )

    if data.get("user_id"):
        stmt = stmt.filter(userReport.id == data["user_id"])
    elif data.get("cathedra_id"):
        stmt = stmt.filter(userReport.cathedra_id == int(data["cathedra_id"]))
    elif data.get("faculty_id"):
        stmt = stmt.filter(model.Cathedras.faculty_id == int(data["faculty_id"]))
    return stmt


@data_reports_app.route("/faculties")
def get_faculties():
    try:
        role = token_utils.get_role()
    except RuntimeError as e:
        return flask.Response(
            str(e),
            status=400,
        )
    try:
        token_utils.check_role(
            [
                db_utils.Role.TEACHER,
                db_utils.Role.HEAD_OF_CATHEDRA,
                db_utils.Role.INSPECTOR,
                db_utils.Role.HEAD_OF_HUMAN_RESOURCES,
            ],
            role,
        )
    except RuntimeError as e:
        return flask.Response(
            str(e),
            status=400,
        )

    with db_utils.auto_session(db_utils.engine_reader) as session:
        stmt = sqlalchemy.select(model.Faculties)

        rows = session.execute(stmt).all()
        mapping = [row._mapping["Faculties"] for row in rows]

        return json.dumps(
            [
                {
                    "id": row.id,
                    "name": row.name,
                }
                for row in mapping
            ]
        )

File: static/js/DataReports.js

import Token from './Token.js'

export default {
    components: {

    },
    data() {
        return {
            report_html_body: "",
            faculties: [],


            cathedras: [],
            teachers:[],
            extend: false,

            filter: {"from":"","till":"","faculty":{},"cathedra":{},"user":{},"extended":false},
            selected_id: "",
        };
    },
    methods: {
        fillReport() {
            let body = this.filter;
            axios.put('/data-reports/' + this.selected_id, body,{
                headers: {
                    'Token': Token.token,
                }
            })
            .then((res) => {
                this.report_html_body = res.data;
            })
            .catch((error) => {
              console.log(error.response.data);
            })
        },
        selectExecutionReport(){
            this.selected_id = 'execution_progress';

            var token_data = Token.getTokenData();
            if(['teacher', 'head_of_cathedra'].includes(token_data.role)){
                this.filter.faculty.id = token_data.faculty_id;
                this.filter.cathedra.id = token_data.cathedra_id;
                this.getTeachers();
            }
            if(['teacher'].includes(token_data.role)){
                this.filter.user.id = token_data.user_id;
            }
        },
        selectSigningReport(){
            this.selected_id = 'signing_progress';

        },
        selectSigningLogReport(){
            this.selected_id = 'signing_log';
        },
        selectDataChangeLogReport(){
            this.selected_id = 'data_change_log';
        },
        getFaculties(){
            axios.get('/faculties', {
                headers: {
                    'Token': Token.token,
                }
            })
            .then((res) => {
                this.faculties = res.data;
                this.faculties.splice(0, 0, {});
            })
            .catch((error) => {
              console.log(error.response.data);
            })
        },
        getCathedras(){
            this.filter.cathedra = {};
            this.filter.user = {};
            this.cathedras = [];
            this.teachers = [];

            if (!this.filter.faculty) {
                return
            }
            axios.get('/cathedras', {
                params:{
                        faculty_id: this.filter.faculty.id,
                },
                headers: {
                    'Token': Token.token,
                }
            })
            .then((res) => {
                this.cathedras = res.data;
                this.cathedras.splice(0, 0, {});
            })
            .catch((error) => {
              console.log(error.response.data);
            })
        },
        getTeachers(){
            this.filter.user = {};
            this.teachers = [];
            if (!this.filter.cathedra || !this.filter.faculty) {
                return
            }
            axios.get('/users', {
                params:{
                        faculty_id: this.filter.faculty.id,
                        cathedra_id : this.filter.cathedra.id,
                        role: 'teacher',
                },
                headers: {
                    'Token': Token.token,
                }
            })
            .then((res) => {
                this.teachers = res.data;
                this.teachers.splice(0, 0, {});
            })
            .catch((error) => {
              console.log(error.response.data);
            })
        },
        showReport(reportName){
            if(['signing_progress', 'signing_log', 'data_change_log'].includes(reportName)){
                var token_data = Token.getTokenData();

                return token_data.role === "head_of_human_resources";
            }
        },
        showFilterItem(item){
            var token_data = Token.getTokenData();
            var role = token_data.role;

            if(this.selected_id == 'signing_progress'){
                return ['period-from', 'period-till', 'faculty', 'cathedras', 'user'].includes(item);
            }else if(this.selected_id == 'execution_progress'){
                if(role == 'teacher'){
                    var bla = ['period-from', 'period-till', 'extended-view'].includes(item);
                    return ['period-from', 'period-till', 'extended-view'].includes(item);
                }else if(role == 'head_of_cathedra'){
                    return ['period-from', 'period-till', 'user', 'extended-view'].includes(item);
                }else if(role == 'head_of_human_resources'){
                    return ['period-from', 'period-till', 'faculty', 'cathedras', 'user', 'extended-view'].includes(item);
                }

            }else if(this.selected_id == 'signing_log' || this.selected_id == 'data_change_log'){
                return ['period-from', 'period-till'].includes(item);

            }

        }
    },
    mounted() {
        this.getFaculties();

    },
    template: `
        <div class="centered-div">
            <table id="data-reports-main-table">
                <tr>
                    <td>
                        <ul>
                            <li v-on:click="selectExecutionReport" v-bind:class="'execution_progress' == selected_id ? 'selected-row':''">execution progress</li>
                            <li v-on:click="selectSigningReport" v-bind:class="'signing_progress' == selected_id ? 'selected-row':''" v-if="showReport('signing_progress')">signing progress</li>
                            <li v-on:click="selectSigningLogReport" v-bind:class="'signing_log' == selected_id ? 'selected-row':''" v-if="showReport('signing_log')">signing log</li>
                            <li v-on:click="selectDataChangeLogReport" v-bind:class="'data_change_log' == selected_id ? 'selected-row':''" v-if="showReport('data_change_log')">data change log</li>
                        </ul>
                    </td>
                    <td>
                       <div>
                            <label for="period-from" v-if="showFilterItem('period-from')">From:</label>
                            <input id="period-from" type="date" v-model="filter.from" v-if="showFilterItem('period-from')">
                            <label for="period-till" v-if="showFilterItem('period-till')">Till:</label>
                            <input id="period-till" type="date" v-model="filter.till" v-if="showFilterItem('period-till')">

                            <label for="faculty" v-if="showFilterItem('faculty')">Faculty:</label>

                            <p-dropdown v-model="filter.faculty" v-bind:options="faculties" filter optionLabel="name" placeholder="Select a Faculty" class="w-full md:w-14rem" v-on:change="getCathedras" v-if="showFilterItem('faculty')">
                                <template #value="slotProps">
                                    <div v-if="slotProps.value" class="flex align-items-center">
                                        <div>{{ slotProps.value.name }}</div>
                                    </div>
                                    <span v-else>
                                        {{ slotProps.placeholder }}
                                    </span>
                                </template>
                                <template #option="slotProps">
                                    <div class="flex align-items-center">
                                        <div>{{ slotProps.option.name }}</div>
                                    </div>
                                </template>
                            </p-dropdown>

                            <label for="cathedras" v-if="showFilterItem('cathedras')">Cathedra:</label>

                            <p-dropdown v-model="filter.cathedra" v-bind:options="cathedras" filter optionLabel="name" placeholder="Select a Cathedra" class="w-full md:w-14rem" v-on:change="getTeachers" v-if="showFilterItem('cathedras')">
                                <template #value="slotProps">
                                    <div v-if="slotProps.value" class="flex align-items-center">
                                        <div>{{ slotProps.value.name }}</div>
                                    </div>
                                    <span v-else>
                                        {{ slotProps.placeholder }}
                                    </span>
                                </template>
                                <template #option="slotProps">
                                    <div class="flex align-items-center">
                                        <div>{{ slotProps.option.name }}</div>
                                    </div>
                                </template>
                            </p-dropdown>


                            <label for="user" v-if="showFilterItem('user')">User:</label>

                            <p-dropdown v-model="filter.user" v-bind:options="teachers" filter optionLabel="full_name" placeholder="Select a User" class="w-full md:w-14rem" v-if="showFilterItem('user')">
                                <template #value="slotProps">
                                    <div v-if="slotProps.value" class="flex align-items-center">
                                        <div>{{ slotProps.value.full_name }}</div>
                                    </div>
                                    <span v-else>
                                        {{ slotProps.placeholder }}
                                    </span>
                                </template>
                                <template #option="slotProps">
                                    <div class="flex align-items-center">
                                        <div>{{ slotProps.option.full_name }}</div>
                                    </div>
                                </template>
                            </p-dropdown>


                            <label for="extended-view" v-if="showFilterItem('extended-view')">Extended view:</label>
                            <input id="extended-view" type="checkbox" v-model="filter.extended" v-if="showFilterItem('extended-view')">

                            <br>
                            <button v-on:click="fillReport">Fill</button>
                        </div>

                        <div id="report-body" v-html="report_html_body">

                        </div>
                    </td>
                </tr>
            </table>
        </div>
    `,
};

